### ISSUE TITLE
Playlist Out of Sync Issue

### ISSUE THREAD

**Issue by awused** [calzoneman/sync#180]
While just watching in my browser the playlist got out of sync. Naoko was properly synchronized, but I saw a different video playing than the single video that was in my playlist. After the video ended and Naoko added more videos I saw video 1 highlighted while video 0 was playing. Refreshing fixed the problem.

I'm not sure how to reproduce it, I'll update this if I figure that out, but it's never happened before.

**Comment by calzoneman** [comment]
Can you tell me more about your setup?  Are the videos being added in advance, or is Naoko adding the next video immediately before the current one ends?

**Comment by awused** [comment]
It does seem to happen when I $skip the only video on the playlist. However that doesn't explain the initial bug I had, where I saw only one video on the playlist and it was the wrong video.

It does seem to be a case of race conditions or frames being handled out of order by the client.

**Comment by calzoneman** [comment]
The only time I've had issues with videos being glitched is when I queue or queue next within a second of the end of a video, and I haven't tested that on v2 yet.

**Comment by awused** [comment]
Unsure but it seems like videos can get added to the playlist out of order, which causes my issue. Both the server and Naoko agree, but my client doesn't.

**Comment by calzoneman** [comment]
I reproduced a playlist glitch by hitting queue next just before a video ended.  I haven't fully investigated, but it's probably due to the animation not being finished when the video changes, so the queue DOM doesn't match the server playlist.

I'm unsure whether it would be worth trying to counteract that or just adding a check to see if the playlist is bugged and refreshing it.

**Comment by calzoneman** [comment]
https://github.com/calzoneman/sync/commit/f10a5d7ec3531c99ca91a0c2d0d16b57fbbddbe1 and https://github.com/calzoneman/sync/commit/a89f5138d3fe9f340d5c70e5b63542f3f7b9cdc0 should solve a lot of cases in which this would happen.

Respectively, they wait until the client's done processing any playlist changes before setting the highlight or adding a new video.

Please let me know if this corrects your issue.

**Comment by awused** [comment]
Issue still exists, may have changed. I need to see it again to confirm it, but I think the client playlist is displaying in the right order now but it's highlighting the wrong one.

I wasn't really paying enough attention but I can still at least confirm the wrong video was highlighted.

**Comment by calzoneman** [comment]
Continued in #182

**Comment by awused** [comment]
This issue should still be open.

Using UIDs will prevent some parts of this on their own, but videos appearing out of order will still be a problem depending on how you implement the UIDs. Also depending on how you do it we could run into the issue where some videos don't appear at all.

Not sure if you're implementation will run into that, but inserting video FFF1 after video FFF0 when FFF0 hasn't finished animating could result in FFF1 just being lost as far as the client is concerned.

**Comment by calzoneman** [comment]
Order is guaranteed with my implementation (currently it provides a uid to insert after).  Animation shouldn't affect inserts because the item is immediately inserted as invisible and then animates to full height.

**Comment by calzoneman** [comment]
Didn't have much time last night, but here's the plan if you're wondering:
- Each playlist item is assigned a unique integer uid
- The client has a mutex for the playlist to ensure only one callback may act on it at a time
- The client maintains a queue of incoming playlist modification packets and executes one at a time, in order (packets are guaranteed to arrive in the order that the server sends them, and the mutex combined with this queue will ensure that they are processed in the correct order)

**Comment by calzoneman** [comment]
If you wouldn't mind, I have a test server up at http://sb.calzoneman.net:8080 (it's an image of the main server as of today)

**Comment by awused** [comment]
`{"media":{"id":"hU1wdTUkdGQ","title":"[HQ] Color Test in Touhou (3/4)","seconds":267,"duration":"04:27","type":"yt","queueby":"","temp":false,"hash":"583e0d5042016f7dc516aa4f5fed718e69cb85f5"},"uid":0,"temp":false,"queueby":"Naoko"}`

Seems a little confused. You've got some information duplicated here (`queueby`, `temp`) and both a `uid` and a `hash`. Also the hash seems logically tied to the specific video and the uid to the playlist entry, which seems odd. I don't see the hash actually being used anywhere, you'll probably want to drop it.

I get two `queue` responses when I add a single video. You might think about renaming the "from" field of `moveMedia` and other similar frames but no one else is even going to notice.

As a side-effect of using uids, when you shuffle you can just send an array of ordered uids and let the client reorder his own list rather than dropping and remaking the entire playlist.

**Comment by calzoneman** [comment]
I decided to re-structure a bit so that a Media object has only media data, and a PlaylistItem object stores media, queueby, temp, and uid.  I have removed the extra variables from Media.

Double queue was because of an if case that was supposed to return but fell through instead (if you noticed, it only double queues stuff that's already cached in the library).

A lot of individual playlist manipulations is generally more expensive than one large manipulation.
